src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2525,39 +2525,95 @@ end
-        else
+        if !haskey(seen_cols, col)
             seen_cols[col] = nothing
-        end
-        # p might be a column of df so we make sure we unalias
-        if col === p
-            p = copy(p)
+            _cycle_permute!(col, cp)
         end
     end
+    
+    return df
+end
 
-    pp = similar(p)
+# convert a classical permutation to zero terminated cycle 
+# notation, zeroing the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
+    firstindex(p) == 1 || 
+        throw(ArgumentError("Permutation vectors must have 1-based indexing"))
+    # this length is sufficient because we do not record 1-cycles,
+    # so the worst case is all 2-cycles.
+    out = similar(p, 3 * length(p) ÷ 2) 
+    out_len = 0
+    start = 0
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start + 1)
+        start isa Int || throw(ArgumentError("Not a permutation"))
+        last_k = p[start]
+        count -= 1
+        last_k == start && continue
+        out_len += 1
+        out[out_len] = last_k
+        p[start] = 0
+        start <= last_k <= length(p) || throw(ArgumentError("Not a permutation"))
+        out_len += 1
+        k = out[out_len] = p[last_k]
+        while true
+            count -= 1
+            p[last_k] = 0
+            last_k = k
+            start <= k <= length(p) || throw(ArgumentError("Not a permutation"))
+            out_len += 1
+            k = out[out_len] = p[k]
+            k == 0 && break
+        end
+        last_k == start  || throw(ArgumentError("Not a permutation"))
+    end
+    return resize!(out, out_len)
+end
 
-    for (i, col) in enumerate(eachcol(df))
-        if !(i in toskip)
-            copyto!(pp, p)
-            fun(col, pp)
+# Permute a vector `v` based on a permutation `p` listed in zero terminated
+# cycle notation. For example, the permutation 1 -> 2, 2 -> 3, 3 -> 1, 4 -> 6,
+# 5 -> 5, 6 -> 4 is traditionaly expressed as [2, 3, 1, 6, 5, 4] but in cycle
+# notation is expressed as [1, 2, 3, 0, 4, 6, 0]
+function _cycle_permute!(v::AbstractVector, p::AbstractVector{<:Integer})
+    i = firstindex(p)
+    @inbounds while i < lastindex(p)
+        last_p_i = p[i]
+        start = v[last_p_i]
+        while true
+            i += 1
+            p_i = p[i]
+            p_i == 0 && break
+            v[last_p_i] = v[p_i]
+            last_p_i = p_i
         end
+        v[last_p_i] = start
+        i += 1
     end
-    return df
+    return v
 end
 
 """
     permute!(df::AbstractDataFrame, p)
 
 Permute data frame `df` in-place, according to permutation `p`.
-No checking is done to verify that `p` is a permutation.
+Throws `ArgumentError` if `p` is not a permutation.
 
-To return a new data frame instead of permuting `df` in-place, use `df[p]`.
-Note that this is generally faster than `permute!(df, p)` for large data frames.
+To return a new data frame instead of permuting `df` in-place, use `df[p, :]`.
```

Comment:
While we are at it can you please also correct line 2621 in this PR (`permute!` should be `invpermute!` there). Thank you

Reply:
Done